### ISSUE TITLE
Fuse -> ROS 2 Rolling Porting Progress

### ISSUE THREAD

**Issue by methylDragon** [locusrobotics/fuse#276]
# Description

This issue tracks the progress of the ROS 2 port for Fuse, targeting the `rolling` release of ROS 2.
Some of the work here will be based off the work mentioned in https://github.com/locusrobotics/fuse/issues/257.

# Architectural Decisions
I will be prioritising API consistency between the ROS 1 and ROS 2 versions of Fuse. This means I won't be looking at high level architectural changes at this time.

# Approach
PRs will be sent against the `rolling` branch on [fuse](https://github.com/locusrobotics/fuse) and [tf2_2d](https://github.com/locusrobotics/tf2_2d), working up the dependency graph and doubling back as necessary. (Or more likely, finding features to port as I work up the dependency graph, and port them for downstream packages as well.)

Changes in dependent packages that have to be made as a result of a change in a dependency packages will be made on a best effort basis, but can't be confirmed until they're buildable, so expect bugs in those PRs.

I fully expect `fuse_core` to take up the bulk of the migration effort. And I also fully expect needing to **circle back** after things finally build and run.

![image](https://user-images.githubusercontent.com/20265670/194437811-f1792ebc-9d9f-47f6-abb9-412b562db46f.png)

# Progress

## General
- [x] Logging https://github.com/locusrobotics/fuse/pull/279
- [x] Resolve macro usage warnings https://github.com/locusrobotics/fuse/pull/280
- [x] Port Time:
  - [x] https://github.com/locusrobotics/fuse/pull/283
  - [x] https://github.com/locusrobotics/fuse/pull/303
- [x] ~~Port messages https://github.com/locusrobotics/fuse/pull/285~~ (DEFERRED to on a per-package basis)
- [x] Linting

BONUS (if time permits)
- (Probably bonus, too many parts to shake up) Port the pluginlib usage to rclcpp_components
- "Dynamic reconfigure" for params

## Down (Up?) The Dependency Tree
- [x] fuse_msgs
	- [x] https://github.com/locusrobotics/fuse/pull/277
- [x] fuse_core
  - [x] https://github.com/locusrobotics/fuse/pull/281
  - [x] https://github.com/locusrobotics/fuse/pull/284 (depends on #283)
  - [x] https://github.com/locusrobotics/fuse/pull/286
  - [x] https://github.com/locusrobotics/fuse/pull/292
  - [x] https://github.com/locusrobotics/fuse/pull/293 (should be merged before the one after)
  - [x] https://github.com/locusrobotics/fuse/pull/294
  - [x] https://github.com/locusrobotics/fuse/pull/285
- [x] fuse_loss:
  - [x] https://github.com/locusrobotics/fuse/pull/287
  - [x] https://github.com/locusrobotics/fuse/pull/295
- [x] fuse_variables:
	- [x] https://github.com/locusrobotics/fuse/pull/288
	- [x] https://github.com/locusrobotics/fuse/pull/296
- [x] fuse_graphs:
	- [x] https://github.com/locusrobotics/fuse/pull/289
	- [x] https://github.com/locusrobotics/fuse/pull/297
- [x] fuse_constraints:
	- [x] https://github.com/locusrobotics/fuse/pull/290
	- [x] https://github.com/locusrobotics/fuse/pull/298
- [x] fuse_publishers
  - [x] https://github.com/locusrobotics/fuse/pull/299
  - [x] https://github.com/locusrobotics/fuse/pull/305
  - [x] https://github.com/locusrobotics/fuse/pull/311
  - [x] https://github.com/locusrobotics/fuse/pull/314
- [x] tf2_2d
	- [x] https://github.com/ros2/geometry2/pull/559
	- [x] https://github.com/locusrobotics/tf2_2d/pull/5 (depends on above)
- [x] fuse_models
  - [x] https://github.com/locusrobotics/fuse/pull/304
  - [x] https://github.com/locusrobotics/fuse/pull/315
- [x] fuse_optimizers
  - [x] https://github.com/locusrobotics/fuse/pull/307
  - [x] https://github.com/locusrobotics/fuse/pull/316
- [x] fuse_viz
  - [x] https://github.com/locusrobotics/fuse/pull/301
  - [x] https://github.com/locusrobotics/fuse/pull/306
  - [x] https://github.com/locusrobotics/fuse/pull/312
- [x] fuse_tutorials
	- [x] https://github.com/locusrobotics/fuse/pull/309
	- [x] https://github.com/locusrobotics/fuse/pull/317
- [x] fuse_doc
  - [x] https://github.com/locusrobotics/fuse/pull/278
- [x] fuse: https://github.com/locusrobotics/fuse/pull/310
- [x] Test fuse_viz when tutorials is done (https://github.com/locusrobotics/fuse/pull/301)

# Additional Notes
- I've tried to migrate away from boost as much as possible (into std), but there aren't equivalents for everything (e.g. serialization).

**Comment by methylDragon** [comment]
Fuse functionality is now ported! :tada: 
@svwilliams 

There's still some linting and docs porting to do

EDIT: Those are DONE!!

**Comment by sloretz** [comment]
@methylDragon Is there anything left to do before closing this ticket?

**Comment by methylDragon** [comment]
I think we're done here! Ticket closed!!
